src/Internal/Hydration/AbstractHydrator.php

```diff
@@ -341,33 +342,54 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
         }
 
         foreach ($this->resultSetMapping()->nestedNewObjectArguments as $objIndex => ['ownerIndex' => $ownerIndex, 'argIndex' => $argIndex]) {
-            if (! isset($rowData['newObjects'][$objIndex])) {
+            if (! isset($rowData['newObjects'][$ownerIndex . ':' . $argIndex])) {
                 continue;
             }
 
-            $newObject = $rowData['newObjects'][$objIndex];
-            unset($rowData['newObjects'][$objIndex]);
+            $newObject = $rowData['newObjects'][$ownerIndex . ':' . $argIndex];
+            unset($rowData['newObjects'][$ownerIndex . ':' . $argIndex]);
 
-            $class = $newObject['class'];
-            $args  = $newObject['args'];
-            $obj   = $class->newInstanceArgs($args);
+            $class          = $newObject['class'];
+            $args           = $newObject['args'];
+            $argColumnNames = $newObject['argColumnNames'];
```

Comment:
I would avoid creating 3 new variables, here and below, but maybe I'm missing something?